<img src = "https://drive.google.com/uc?export=view&id=1WaM3ez8iLaUk3VyWNYZQuifnvbEX4vbK" alt = "Encabezado MLDS" width = "100%">  </img>


#**Quiz 3 : Entendimiento de los datos con *pandas***
---

Este es un quiz para practicar las habilidades adquiridas usando *pandas* en el proceso de exploración y entendimiento de conjuntos de datos. Utilizaremos el [dataset Titanic](https://www.kaggle.com/c/titanic) disponible en *Kaggle*.

El conjunto de datos del Titanic cuenta con múltiples variables de la tragedia:

| Variable | Definición	| Valores |
| --- | --- | --- |
| survival | 	Supervivencia  | 	0 = No, 1 = Sí |
| pclass 	| Clase del tiquete | 	1 = 1ra, 2 = 2da, 3 = 3ra
| sex 	| Sexo 	| |
| Age |	Edad en años 	| |
| sibsp |	# de hermanos / cónyuge abordo del Titanic 	| |
| parch |	# de padres / hijos abordo del Titanic 	| |
| ticket |	Número del ticket | |
| fare 	| Costo del ticket | |
| cabin |	Número de la cabina 	| |
| embarked |	Puerto de embarque |	C = Cherbourg, Q = Queenstown, S = Southampton |

En este quiz usted debe realizar un proceso de análisis exploratorio de datos con *Pandas* con el objetivo de practicar los conceptos y sus habilidades del entendimiento de los datos.

> **Nota:** Esta tarea va a ser calificada en la plataforma **[UNCode](https://juezun.github.io/)**. Para esto, en cada ejercicio se indicará si es calificable o no, también los lugares donde debe escribir su código sin modificar lo demás con un aproximado de cantidad de líneas a escribir. No se preocupe si su código toma más líneas, esto es simplemente un aproximado destinado a que pueda replantear su estrategia si el código está tomando más de las esperadas. No es un requisito estricto y soluciones más largas también son válidas. Al finalizar, para realizar el envío (*submission*), descargue el notebook como un archivo **`.ipynb`** y haga su entrega a través de la plataforma de aprendizaje.

Ejecute la siguiente celda para importar *pandas*.

In [41]:
# Importar pandas
import pandas as pd
import numpy as np

In [2]:
#TEST_CELL
!python --version
print('Pandas', pd.__version__)

Python 3.11.5


Pandas 2.1.0


Este material fue realizado con las siguientes versiones:

- Python 3.10.6
- pandas 1.5.3

## **0.  Leer el archivo de datos**
---
> **IMPORTANTE:**

  1. Por favor descargue el archivo con el conjunto de datos desde el siguiente enlace: **[titanic.csv](https://raw.githubusercontent.com/JuezUN/datasets/master/titanic.csv)**.
  
  2. En el enlace, haga clic derecho y seleccione la opción **"Guardar como..."**.

  3. Cargue este archivo en su sistema de archivos de **Google Colaboratory**. Para hacer esto:
    
    * Haga clic donde aparece el icono del directorio en la parte izquierda de la ventana en *Google Colaboratory* (resaltado en azul en la siguiente imagen con el número 1).
    * Después, haga clic en el icono **"Subir"** (resaltado en azul en la siguiente imagen con el número 2).
    * Seleccione el archivo **`titanic.csv`** que previamente descargó y haga clic en **"Aceptar"**.

<img src = "https://drive.google.com/uc?export=view&id=13W59jfdOpAn_DYIjpXt4WPBZiU7BlRGY" alt = "Sistema archivos colab" width = "70%">  </img>

4. Ejecute la siguiente celda para cargar el conjunto de datos dentro del DataFrame de *pandas* llamado **`df`**:

In [4]:
### EJECUTAR ESTA CELDA SIN MODIFICARLA. SI FALLA, DEBE SEGUIR EL PROCEDIMIENTO DESCRITO ANTES.
df = pd.read_csv('titanic.csv')

No se preocupe por el archivo del dataset en ***UNCode***, el ambiente de calificación ya cuenta con este archivo para calificar sus ejercicios.

## **1. Tipos de variables**
---
Usted deberá identificar y explorar los diferentes tipos de variables (columnas) del conjunto de datos y guardar las respuestas solicitadas en las variables solicitadas en cada punto.

Utilice las siguientes celdas para explorar libremente el conjunto de datos utilizando *pandas*. Después, consigne sus respuestas en las preguntas que aparece más adelante.


In [6]:
# Wide glimpse of the data frame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
# Summary-principal descriptive statistics
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [5]:
# Five random data frame values
df.sample(5)

PassengerId  Survived  Pclass  \
27            28         0       1   
431          432         1       3   
533          534         1       3   
149          150         0       2   
203          204         0       3   

                                                  Name     Sex   Age  SibSp  \
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
431  Thorneycroft, Mrs. Percival (Florence Kate White)  female   NaN      1   
533             Peter, Mrs. Catherine (Catherine Rizk)  female   NaN      0   
149                  Byles, Rev. Thomas Roussel Davids    male  42.0      0   
203                               Youseff, Mr. Gerious    male  45.5      0   

     Parch  Ticket      Fare        Cabin Embarked  
27       2   19950  263.0000  C23 C25 C27        S  
431      0  376564   16.1000          NaN        S  
533      2    2668   22.3583          NaN        C  
149      0  244310   13.0000          NaN        S  
203      0    2628    7.2250          NaN        C

### **1.1.  Variables con valores faltantes**
---

> **¿Qué variables tienen filas con valores faltantes?**

Llene la lista **`valores_faltantes`** con la respuesta correcta. El orden en que ponga los nombres de las variables no es importante.

La respuesta debería ser con un subconjunto de los nombres de las variables del conjunto de datos:

**`['PassengerId', 'Survived', 'Pclass',
       'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare',
       'Cabin', 'Embarked']`**


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Pistas</b></font>
</summary>

* En este ejercicio deberá llenar la lista de manera manual. Para esto, puede apoyarse en las funciones de exploración de los datos como **`.describe`** o **`.info`** para identificar las variables con valores faltantes.

* Considere el total de registros en el conjunto de datos y el total de registros con valores no-nulos por cada variable para encontrar las diferencias.


In [35]:
# Empty values
valores_faltantes = []
# Number of rows
rows = df.shape[0]
# Iterates over data frame columns
for column in df.columns:
    # If specific column rows are less than the df row, then empty values
    if rows > df[column].count():
        valores_faltantes.append(column)

valores_faltantes

['Age', 'Cabin', 'Embarked']

### **1.2.  Variables categóricas**
---

> **¿Qué variables tienen valores categóricos?**

Llene la lista **`categoricas`** con la respuesta correcta. El orden en que ponga los nombres de las variables no es importante.

La respuesta debería ser un subconjunto de todas las variables del conjunto de datos:

**`['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']`**


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Pistas</b></font>
</summary>

* En este ejercicio deberá llenar la lista de manera manual. Para esto, puede apoyarse en las funciones de exploración de los datos para identificar las variables con valores categóricos. Por ejemplo: la columna **'Sex'** puede tener únicamente los valores **'female'** y **'male'**.

* Considere como valores categóricos aquellos que puedan representar características del pasajero o grupos a los que pertenezca.

* Aunque la variable sea cargada como un número, es posible que represente una variable categórica.

In [43]:
# VARIABLE CALIFICADA categóricas

# Query for all category variables, then pass it to a python list
categoricas = (
    df
    .astype(
        {
            'Sex': 'category',
            'Survived': 'category',
            'Pclass': 'category',
            'Embarked': 'category',
        }
    )
    .select_dtypes(include=['category'])
    .columns
    .tolist()
)

categoricas


['Survived', 'Sex', 'Cabin', 'Embarked']

### **1.3.  Conteos de embarques**
---

> **¿Cuántos pasajeros embarcaron en cada uno de los puertos?**

Llene la lista **`embarcaron`** con la respuesta correcta. El orden en que ponga los números de pasajeros que embarcaron en cada uno de los puertos no es importante.

La respuesta debería debe tener 3 valores numéricos correspondientes al número de personas que embarcaron en:
* *Southampton* (S)
*  *Cherbourg* (C)
* *Queenstown* (Q)


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Pistas</b></font>
</summary>

* En este ejercicio deberá llenar la lista de manera manual. Debe ayudarse para encontrar los resultados usando las funciones de explicación de datos. Recuerde que *pandas* proporciona una función que realiza los conteos de los datos categóricos almacenados en una variable.

* Recuerde que la columna **'Embarked'** almacena los datos acerca del puerto de embarque de los pasajeros.

In [25]:
# VARIABLE CALIFICADA embarcaron:

# Gets how many people boarded at ship ports
embarcaron = (
    df
    .Embarked
    .value_counts()
    .tolist()
)

## **2. Supervivencia**
---
El conjunto de datos *Titanic* cuenta con la variable **`Survived`**. Esta indica quiénes sobrevivieron a la tragedia, representando a los sobrevivientes con valor igual a $1$ y las víctimas con el valor $0$.

Nos interesa saber qué variables influyeron más en la supervivencia de una persona.

* Por cada variable numérica encuentre el valor medio de aquellos que sobrevivieron.

* En el caso de la variable categórica **`Sex`**, encuentre los valores de la proporción de sobrevivientes para mujeres (*female*) y hombres (*male*).



### **2.1. Supervivencia en variables numéricas**
---

Encontrar el valor medio de todas las variables numéricas para aquellos que sobrevivieron.

Aunque es posible obtener el resultado en una sola línea de código, se recomienda primero calcular los promedios de las variables numéricas para ambos grupos de pasajeros: los que sobrevivieron y los que no, y posteriormente, seleccionar los promedios de los que sobrevivieron. Por lo anterior, este ejercicio se divide en dos partes.

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Pistas</b></font>
</summary>

* Recuerde que puede conformar grupos a partir de variables o condiciones con el uso de la función **`groupby`** para después realizar operaciones de agregación como el promedio **`mean`** o la suma **`sum`**.
* Para este ejercicio deberá considerar las variables con valores numéricos, incluso si son interpretadas como variables categóricas. Delegue esta decisión a *pandas* y sus funciones que sólo operan con valores numéricos.


Primero, encuentre los promedios de las variables numéricas para ambos grupos de pasajeros: sobrevivientes y no sobrevivientes.
> **Nota**: Tenga en cuenta que el tipo de dato que se solicita es un objeto ***Dataframe*** llamado **'promedios_por_grupo'**.

In [45]:
# VARIABLE CALIFICADA promedios_por_grupo:

# Group by Survived and gets the mean, only for numeric values
promedios_por_grupo = (
    df
    .groupby('Survived')
    .mean(numeric_only=True)
)
promedios_por_grupo

PassengerId    Pclass        Age     SibSp     Parch       Fare
Survived                                                                 
0          447.016393  2.531876  30.626179  0.553734  0.329690  22.117887
1          444.368421  1.950292  28.343690  0.473684  0.464912  48.395408

In [31]:
#TEST_CELL
promedios_por_grupo > 10

PassengerId  Pclass   Age  SibSp  Parch  Fare
Survived                                               
0                True   False  True  False  False  True
1                True   False  True  False  False  True

**Salida esperada:**

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PassengerId</th>
      <th>Pclass</th>
      <th>Age</th>
      <th>SibSp</th>
      <th>Parch</th>
      <th>Fare</th>
    </tr>
    <tr>
      <th>Survived</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>True</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>True</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>True</td>
    </tr>
  </tbody>
</table>

Ahora debe seleccionar los promedios del grupo que sobrevivió. Por ejemplo, para aquellos que sobrevivieron el promedio de edad (variable **`Age`**) es aproximadamente **`28.343690`** años.

Al interpretar los resultados puede ignorar la variable **`PassengerId`** que, aunque es un valor numérico, es usada como identificador. En cambio, podría ser interesante comparar las diferencias en las tarifas (**`Fare`**) entre ambos grupos.

Deje los promedios obtenidos por variable en un objeto *Series* de *pandas* llamado **`supervivencia_numericas`**.


> **Nota:** Tenga presente que el tipo de dato que se solicita es un objeto ***Series***.

In [51]:
# VARIABLE CALIFICADA supervivencia_numericas:

# Querys for people who survived and gets the mean of the numeric values
supervivencia_numericas = (
    df
    .query("Survived == 1")
    .drop(columns='Survived')
    .mean(numeric_only=True)
)

supervivencia_numericas = pd.Series(supervivencia_numericas, name=1)

In [52]:
#TEST_CELL
supervivencia_numericas.astype(int)

PassengerId    444
Pclass           1
Age             28
SibSp            0
Parch            0
Fare            48
Name: 1, dtype: int64

**Salida esperada:**
```python
PassengerId    444
Pclass           1
Age             28
SibSp            0
Parch            0
Fare            48
Name: 1, dtype: int64
```

### **2.2. Supervivencia en variables categóricas**
---
Construya un objeto ***DataFrame*** de *pandas* llamado **`supervivencia_sexo`** que contenga el promedio de supervivencia en cada grupo de la variable **`Sex`**. Debe encontrar los valores de las proporciones de supervivientes para mujeres y hombres por separado. Por ejemplo, si hubieran $3$ mujeres sobrevivientes de un total de $4$ mujeres, y $2$ hombres sobrevivientes de un total de $4$ hombres; las proporciones correspondientes serían: **`female: 0.75`** y **`male: 0.5`**.

Puede utilizar la siguiente celda para encontrar los valores correspondientes y ponerlos en el ***DataFrame*** **`supervivencia_sexo`**.

> **Nota:** Tenga presente que el tipo de dato que se solicita es un ***DataFrame***.


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Pistas</b></font>
</summary>

* Recuerde que puede conformar grupos a partir de variables o condiciones con el uso de la función **`groupby`** para después realizar operaciones de agregación como el promedio **`mean`** o la suma **`sum`**.
* La proporción de una variable de valores binarios ($0$ o $1$) puede ser obtenida a partir del promedio, suponiendo los valores en $1$ como las instancias que hacen parte de la proporción, y los que están en $0$ las que no.


In [44]:
# VARIABLE CALIFICADA supervivencia_sexo:

# Divide the group by sex and the add the mean to the survived column
supervivencia_sexo = (
    df
    .groupby('Sex')
    .agg({'Survived': 'mean'})
)

Survived
Sex             
female  0.742038
male    0.188908

In [45]:
#TEST_CELL
(supervivencia_sexo * 100).astype(int)

Survived
Sex             
female        74
male          18

**Salida esperada:**

```python
	    Survived
Sex
female  74
male    18
```

¿Cómo se pueden interpretar estos resultados? Por favor, siempre recuerde analizar e interpretar sus hallazgos.

**¡Felicitaciones!** Ha terminado el quiz. ¡Buen trabajo!

## **Entrega**

Para entregar el notebook por favor haga lo siguiente:
1. Descargue el notebook (`Archivo` -> `Descargar .ipynb`).
2. Ingrese a la plataforma de aprendizaje.
3. Realice el envío del *notebook* que descargó en la tarea (o quiz) correspondiente.
4. Recuerde que si tiene algún error, puede hacer múltiples intentos de envío.

## **Créditos**


* **Profesor:** [Felipe Restrepo Calle](https://dis.unal.edu.co/~ferestrepoca/)
* **Asistentes docentes:**
  - Alberto Nicolai Romero Martínez
  - Miguel Angel Ortiz Marín

**Universidad Nacional de Colombia** - *Facultad de Ingeniería*
